File: python/24_swapNodesInPairs.py
class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if not head or not head.next: return head
        
        prev, curr = head, head.next
        head = head.next
        while prev and prev.next:
            after = curr.next
            prev.next, curr.next = after.next if after and after.next else after, prev
            curr, prev = prev.next, after
        
        return head

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        current = head
        temp = head.next
        current.next = self.swapPairs(temp.next)
        temp.next = current
        return temp

File: swift/24_SwapNodesInPairs.swift
class Solution {
    func swapPairs(_ head: ListNode?) -> ListNode? {
        if head == nil || head?.next == nil { return head }
        
        var prev: ListNode? = head
        var curr: ListNode? = head?.next
        
        prev?.next = swapPairs(curr?.next)
        curr?.next = prev
        
        return curr
    }
}
